File: python/longestConsecutiveSequence.py
# link = https://leetcode.com/problems/longest-consecutive-sequence
class Solution:
    def longestConsecutive(self, nums: list[int]) -> int:
        if not len(nums):
            return 0
        best = 0

        fringe = set(nums)

        for n in fringe:
            # start at smallest and count up
            if (n - 1) in fringe:
                continue  # we aren't the smallest of the sequence
            # otherwise we are... so how long is the sequence?
            cn = n + 1
            length = 1
            while cn in fringe:
                length += 1
                cn += 1
            best = max(best, length)

        return best

File: cpp/longestConsecutiveSequence.cc
#include <bits/stdc++.h>

class Solution {
 public:
  int longestConsecutive(std::vector<int>& nums) {
    std::unordered_set<int> fringe(nums.begin(), nums.end());
    int best{0};

    for (auto&& n : nums) {
      // make sure it exists in our fringe
      auto f = fringe.find(n);
      if (f == fringe.end()) continue;
      fringe.erase(f);  // remove ourselves

      // if it exists, count sequence below and sequence above
      int len{1};  // include ourselves to begin with
      int dec{1}, inc{1};

      while (fringe.find(n - dec) != fringe.end()) {
        // remove from our fringe and increase length
        fringe.erase(n - dec);
        ++len;
        ++dec;
      }

      while (fringe.find(n + inc) != fringe.end()) {
        fringe.erase(n + inc);
        ++len;
        ++inc;
      }

      best = std::max(best, len);
    }

    return best;
  }
};
